### ISSUE TITLE
compressing and decompressing in the same time

### ISSUE THREAD

**Issue by DavidBrenner3** [lz4/lz4#321]
Hi.

First, thank you for this wonderful code.

In one of my apps I use lz4 to lower the memory usage.
The data is stored in chunks.
A thread is decompressing a chunk while another thread is compressing another chunk.
But memory is corrupted sometimes.
If I input a delay between compression and decompression the problem disappears.

Is there a way to fix this problem?

Thank you in advance.

Regards, David

**Comment by Cyan4973** [comment]
This _feels_ like a potential race condition.
Did you test your program with thread sanitizer ?

**Comment by DavidBrenner3** [comment]
Thank you for your quick reply.

Well, the threads are not using the same memory area.
But I can't test with thread sanitizer since most of my app is made in Delphi,
I use lz4 as a dll (compiled in VS2010).

What is strange I use also libjpeg turbo as a dll in multiple threads but don't have problems.

**Comment by Cyan4973** [comment]
The API is multi-threadable,
meaning it shares no information when multiple methods are called from different threads.

I can imagine a few corner-case scenarios where things could go wrong,
but it depends on which exact prototypes are being used and how.
That's the kind of problem thread sanitizer would be able to spot, but unfortunately, it seems not an option in your case.

Seeing the code seems required to debug further.

**Comment by DavidBrenner3** [comment]
If I do this:

            EnterCriticalSection(FCS);
            LZ4_decompress_safe(...);
            LeaveCriticalSection(FCS);
.....
.....
            EnterCriticalSection(FCS);
            LZ4_compress(...);
            LeaveCriticalSection(FCS);

seems ok but the speed drops.

But if I do this:

            EnterCriticalSection(FCS);
            LZ4_decompress_safe(...);
            LeaveCriticalSection(FCS);
.....
.....
            //local variable arrTemp
            LZ4_compress(...arrTemp...);
            EnterCriticalSection(FCS);
            //here I copy arrTemp to main data
            LeaveCriticalSection(FCS);
            FreeAndNil(arrTemp);

speed is ok but the problem reappears.

**Comment by Cyan4973** [comment]
Assuming your application uses _exactly_ the names prototypes `LZ4_decompress_safe()` and `LZ4_compress()`, then these methods are totally thread-safe, and there is no scenario where they could fail.

It means, any corruption is likely happening in the input / output buffer.
Your application seems designed to avoid that. But maybe it nonetheless happen.
Difficult to know, samples are incompletely and do not show it.

If I were you, I would keep the 1st design,
but declare more critical sections.
It seems you are using a single critical section for all operations.
I would recommend to use one per buffer (if it makes sense).
This way, compression and decompression could still run in parallel, as long as they operate on different buffers.

**Comment by DavidBrenner3** [comment]
Thank you for advice.

BTW here is the code + binary: https://drive.google.com/open?id=0ByKxAD_t9uvLRjU2ZW91dTIxMzg
Check the About section...

But the code is too complex to debug the problem...

**Comment by Cyan4973** [comment]
I tried to find where the calls are made.
Your code base is vast, so I couldn't be sure were are the correct calls.
I found a lot of them in `uMain.pas`.

In your example : 
```
            //local variable arrTemp
            LZ4_compress(...arrTemp...);
            EnterCriticalSection(FCS);
            //here I copy arrTemp to main data
            LeaveCriticalSection(FCS);
            FreeAndNil(arrTemp);
```

`arrTemps` is mentioned as a temporary destination buffer,
but no mention is provided regarding the source buffer.

But my understanding of the provided code is that source can also be a cell.
In which case, how can you be guaranteed that it will not be overwritten (by another thread) while compression is ongoing ?

**Comment by DavidBrenner3** [comment]
Found out why the problem appears: it seems that sometimes the threads are not suspended by the Suspend command.
When the user wants to modify the frames the code generates NrOfProcessors - 1 threads, divides the work into NrOfProcessors jobs and start them simultaneously (one in main thread and the rest in the other threads).
When a thread is finishing its job is suspending itself so other job can be assigned. At least it's suppose to.
It says is suspended (the internal FSuspended is True) but it's not. And the code is assigning the input buffer while the thread is doing its job. xDxD

I have no idea why the suspend command not working sometimes.
For now I used this:

      while not Terminated do
      begin
         Suspended := True;
         if (not FJobToDo) and (not Terminated) then //added now
            Continue;
         if Terminated and (not FJobToDo) then
            Break;
         try
            //do work
         finally
            FJobToDo := False;
         end;
      end;
